src/util/amount.rs

```diff
@@ -0,0 +1,367 @@
+#![allow(missing_docs)]
+
+use std::error;
+use std::fmt;
+use std::ops;
+use std::str::FromStr;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Denomination {
+    Bitcoin = 8,
+    MilliBitcoin = 5,
+    Bit = 2,
+    Satoshi = 0,
+    MilliSatoshi = -3,
+}
+
+impl fmt::Display for Denomination {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(match self {
+            Denomination::Bitcoin => "BTC",
+            Denomination::MilliBitcoin => "mBTC",
+            Denomination::Bit => "bits",
+            Denomination::Satoshi => "satoshi",
+            Denomination::MilliSatoshi => "msat",
+        })
+    }
+}
+
+pub trait Precision: Default + Clone {
```

Comment:
I guess `Default + Clone` are needed because the structs get instantiated in Amount, but the better solution would be to use `PhantomData`. Then you wouldn't need to derive these traits for the precision structs.

Reply:
Yeah I already mentioned we could use PhantomData better.